File: app/src/main/java/com/github/pockethub/android/ui/issue/DashboardIssueFragment.java
/*
 * Copyright (c) 2015 PocketHub
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.pockethub.android.ui.issue;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.view.View;

import com.github.pockethub.android.R;
import com.github.pockethub.android.core.issue.IssueStore;
import com.github.pockethub.android.ui.PagedItemFragment;
import com.github.pockethub.android.ui.item.issue.IssueDashboardItem;
import com.github.pockethub.android.ui.item.issue.IssueItem;
import com.github.pockethub.android.util.AvatarLoader;
import com.meisolsson.githubsdk.core.ServiceGenerator;
import com.meisolsson.githubsdk.model.Issue;
import com.meisolsson.githubsdk.model.Page;
import com.meisolsson.githubsdk.service.issues.IssueService;
import com.xwray.groupie.Item;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Map;

import javax.inject.Inject;

import io.reactivex.Single;
import retrofit2.Response;

import static com.github.pockethub.android.RequestCodes.ISSUE_VIEW;

/**
 * Fragment to display a pageable list of dashboard issues
 */
public class DashboardIssueFragment extends PagedItemFragment<Issue> {

    /**
     * Filter data argument
     */
    public static final String ARG_FILTER = "filter";

    private IssueService service = ServiceGenerator.createService(getActivity(), IssueService.class);

    @Inject
    protected IssueStore store;

    @Inject
    protected AvatarLoader avatars;

    private Map<String, Object> filterData;

    @SuppressWarnings("unchecked")
    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        filterData = (Map<String, Object>) getArguments().getSerializable(ARG_FILTER);
        super.onActivityCreated(savedInstanceState);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == ISSUE_VIEW) {
            notifyDataSetChanged();
            forceRefresh();
            return;
        }

        super.onActivityResult(requestCode, resultCode, data);
    }

    @Override
    protected Single<Response<Page<Issue>>> loadData(int page) {
        return service.getIssues(filterData, page);
    }

    @Override
    public void onItemClick(@NonNull Item clickedItem, @NonNull View view) {
        if (clickedItem instanceof IssueDashboardItem) {
            int position = getListAdapter().getAdapterPosition(clickedItem);
            Collection<Issue> issues = new ArrayList<>();
            for (Item item : items) {
                if (item instanceof IssueDashboardItem) {
                    issues.add(((IssueItem) item).getData());
                }
            }
            startActivityForResult(IssuesViewActivity.createIntent(issues, position), ISSUE_VIEW);
        }
    }

    @Override
    protected int getLoadingMessage() {
        return R.string.loading_issues;
    }

    @Override
    protected int getErrorMessage() {
        return R.string.error_issues_load;
    }

    @Override
    protected Item createItem(Issue item) {
        return new IssueDashboardItem(avatars, item);
    }
}

File: app/src/main/java/com/github/pockethub/android/ui/issue/DashboardIssueFragment.kt
/*
 * Copyright (c) 2015 PocketHub
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.pockethub.android.ui.issue

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.github.pockethub.android.ItemListHandler
import com.github.pockethub.android.PagedListFetcher
import com.github.pockethub.android.PagedScrollListener
import com.github.pockethub.android.R
import com.github.pockethub.android.RequestCodes.ISSUE_VIEW
import com.github.pockethub.android.core.issue.IssueStore
import com.github.pockethub.android.ui.base.BaseFragment
import com.github.pockethub.android.ui.item.issue.IssueDashboardItem
import com.github.pockethub.android.ui.item.issue.IssueItem
import com.github.pockethub.android.util.AvatarLoader
import com.github.pockethub.android.util.ToastUtils
import com.meisolsson.githubsdk.model.Issue
import com.meisolsson.githubsdk.model.Page
import com.meisolsson.githubsdk.service.issues.IssueService
import com.xwray.groupie.Item
import com.xwray.groupie.OnItemClickListener
import io.reactivex.Single
import kotlinx.android.synthetic.main.fragment_item_list.view.*
import retrofit2.Response
import java.util.ArrayList
import javax.inject.Inject

/**
 * Fragment to display a pageable list of dashboard issues
 */
class DashboardIssueFragment : BaseFragment() {

    @Inject
    protected lateinit var service: IssueService

    @Inject
    protected lateinit var store: IssueStore

    @Inject
    protected lateinit var avatars: AvatarLoader

    private lateinit var pagedListFetcher: PagedListFetcher<Issue>

    private lateinit var itemListHandler: ItemListHandler

    private lateinit var pagedScrollListener: PagedScrollListener

    private var filterData: Map<String, Any>? = null

    protected val loadingMessage: Int
        get() = R.string.loading_issues

    protected val errorMessage: Int
        get() = R.string.error_issues_load

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        filterData = arguments!!.getSerializable(ARG_FILTER) as Map<String, Any>
        super.onActivityCreated(savedInstanceState)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(R.layout.fragment_item_list, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        itemListHandler = ItemListHandler(
            view.list,
            view.empty,
            lifecycle,
            activity,
            OnItemClickListener(this::onItemClick)
        )

        pagedListFetcher = PagedListFetcher(
            view.swipe_item,
            lifecycle,
            itemListHandler,
            { t -> ToastUtils.show(activity, errorMessage)},
            this::loadData,
            this::createItem
        )

        pagedScrollListener = PagedScrollListener(
            itemListHandler.mainSection,
            pagedListFetcher,
            view.list,
            loadingMessage
        )
        itemListHandler.setEmptyText(R.string.no_repositories)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == ISSUE_VIEW) {
            pagedListFetcher.refresh()
            return
        }

        super.onActivityResult(requestCode, resultCode, data)
    }

    private fun loadData(page: Int): Single<Response<Page<Issue>>> {
        return service.getIssues(filterData, page.toLong())
    }

    private fun onItemClick(clickedItem: Item<*>, view: View) {
        if (clickedItem is IssueDashboardItem) {
            val position = itemListHandler.getItemPosition(clickedItem)
            val issues = ArrayList<Issue>()
            for (item in itemListHandler.items) {
                if (item is IssueDashboardItem) {
                    issues.add((item as IssueItem).issue)
                }
            }
            startActivityForResult(IssuesViewActivity.createIntent(issues, position), ISSUE_VIEW)
        }
    }

    private fun createItem(item: Issue): Item<*> {
        return IssueDashboardItem(avatars, item)
    }

    companion object {

        /**
         * Filter data argument
         */
        val ARG_FILTER = "filter"
    }
}
